scheduler/src/cook/mesos/task.clj

```diff
@@ -0,0 +1,174 @@
+        ports-needed))
+
+(defn add-ports-to-task-info
+  "Given a set of tasks and offers that were matched together by Fenzo,
+   assigns the specific ports requested by role to each task.
+   Returns the input tasks, decorated with :ports-resource-messages"
+  [available-resources tasks]
+  (let [available-ports (available-resources "ports")]
+    (map (fn [task]
+           (let [ports (:ports-assigned task)
+                 port-env-vars (into {} (map-indexed (fn [i p] [(str "PORT" i) (str p)])
+                                                     ports))]
+             (-> task
+                 (assoc :ports-resource-messages (take-ports available-ports ports))
+                 (update-in [:command :environment] merge port-env-vars))))
+         tasks)))
+
+(defn take-resources
+  "Given a set of available resources (in the format returned by resources-by-role),
+  take the specified amount of the specified resource from the pool.
+  Return {:remaining-resources (the pool after resources were taken)
+          :mesos-messages (the Mesos messages that are necessary to reserve the resources)
+          :amount-still-needed (the amount still needed to satisfy amount... if it's not 0, something went wrong)}"
+  [available-resources resource-name amount]
+  (let [avail (available-resources resource-name)
+        ;; to be polite to other frameworks, take from role-specific pool first.
+        sorted-roles (sort-by #(= "*" %) (keys avail))
+        init-state {:remaining-resources avail
+                    :mesos-messages []
+                    :amount-still-needed amount}]
+    (reduce (fn [{:keys [remaining-resources mesos-messages amount-still-needed]
+                  :as state} role-name]
+              (let [amount-avail (or (remaining-resources role-name) 0)
+                    amount-to-take (min amount-still-needed amount-avail)]
+                (if (pos? amount-to-take)
+                  {:remaining-resources
+                   (assoc remaining-resources role-name (- amount-avail amount-to-take))
+                   :mesos-messages
+                   (conj mesos-messages {:name resource-name
+                                         :type :value-scalar
+                                         :role role-name
+                                         :scalar amount-to-take})
+                   :amount-still-needed (- amount-still-needed amount-to-take)}
+                  state)))
+            init-state
+            sorted-roles)))
+
+(defn take-all-scalar-resources-for-task
+  [resources task]
+  (reduce
+   (fn [{:keys [remaining-resources mesos-messages]} [resource-keyword amount]]
+     (let [resource-name (name resource-keyword)
+           adjustment (take-resources remaining-resources resource-name amount)]
+       {:remaining-resources (assoc remaining-resources resource-name (:remaining-resources adjustment))
+        :mesos-messages (vec (concat mesos-messages (:mesos-messages adjustment)))}))
```

Comment:
Wow, I hadn't seen that!  Very interesting.  I just replaced all instances of concat with into.